rfcs/GraphQLOverWebSocket.md

```diff
@@ -0,0 +1,215 @@
+# GraphQL over WebSocket Protocol
+
+## Nomenclature
+
+- **Socket** is the main WebSocket communication channel between the _server_ and the _client_
+- **Connection** is a connection **within the established socket** describing a "connection" through which the operation requests will be communicated
+
+## Communication
+
+The WebSocket sub-protocol for this specification is: `graphql-transport-ws`
+
+Messages are represented through the JSON structure and are stringified before being sent over the network. They are bidirectional, meaning both the server and the client conform to the specified message structure.
+
+**All** messages contain the `type` field outlining the type or action this message describes. Depending on the type, the message can contain two more _optional_ fields:
+
+- `id` used for uniquely identifying server responses and connecting them with the client requests
+- `payload` holding the extra "payload" information to go with the specific message type
+
+The server can terminate the socket (kick the client off) at any time. The close event dispatched by the server is used to describe the fatal error to the client.
+
+The client terminates the socket and closes the connection by dispatching a `1000: Normal Closure` close event to the server indicating a normal closure.
+
+## Keep-Alive
+
+The server will occasionally check if the client is still "alive", available and listening. In order to perform this check, implementation leverages the standardized [Pings and Pongs: The Heartbeat of WebSockets](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#Pings_and_Pongs_The_Heartbeat_of_WebSockets).
+
+Keep-Alive interval and the "pong wait" timeout can be tuned by using the accompanying configuration parameter on the server.
+
+Ping and Pong feature is a mandatory requirement by [The WebSocket Protocol](https://tools.ietf.org/html/rfc6455#section-5.5.2). All clients that don't support it are **not** RFC6455 compliant and will simply have their socket terminated after the pong wait has passed.
+
+## Message types
+
+### `ConnectionInit`
+
+Direction: **Client -> Server**
+
+Indicates that the client wants to establish a connection within the existing socket. This connection is **not** the actual WebSocket communication channel, but is rather a frame within it asking the server to allow future subscription operation requests.
+
+The client can specify additional `connectionParams` which are sent through the `payload` field in the outgoing message.
+
+The server must receive the connection initialisation message within the allowed waiting time specified in the `connectionInitWaitTimeout` parameter during the server setup. If the client does not request a connection within the allowed timeout, the server will terminate the socket with the close event: `4408: Connection initialisation timeout`.
+
```

Comment:
We should require that `ConnectionInit` must only be specified once, otherwise a client could trigger partial reinitialising of the socket which could lead to an undefined state.

```suggestion

If the server receives more than one `ConnectionInit` message, the server will terminate the socket with the close event `...`
```

Reply:
Lets simply disallow multiple `ConnectionInit` requests at any given time 👉 https://github.com/graphql/graphql-over-http/pull/140/commits/ad1ee1f7d709916a13b8a59079f8056c7eef0f43.

Implemented straight away 👉 https://github.com/enisdenjo/graphql-transport-ws/commit/757c6e966ffa1756cca2687b0275d7d7eff2ce87.

Reply:
🎉 Released in [graphql-transport-ws@v1.4.2](https://github.com/enisdenjo/graphql-transport-ws/releases/tag/v1.4.2).